runtime/dynlink.c

```diff
@@ -78,65 +78,155 @@ static c_primitive lookup_primitive(const char * name)
   return NULL;
 }
 
+#endif /* NATIVE_CODE */
+
 /* Parse the ld.conf file and add the directories
    listed there to the search path */
 
 #define LD_CONF_NAME T("ld.conf")
 
-CAMLexport const char_os * caml_get_stdlib_location(void)
+/* Return a copy of [path], interpreting explicit-relative paths relative to
+   [root]. [root] must not end with a directory separator. The result of this
+   function can never be ".", ".." or a path beginning "./" or "../". Note that
+   the function does not necessary canonicalise the path. */
+static char_os *make_relative_path_absolute(char_os *path, char_os *root)
 {
-  const char_os * stdlib;
-  stdlib = caml_secure_getenv(T("OCAMLLIB"));
-  if (stdlib == NULL) stdlib = caml_secure_getenv(T("CAMLLIB"));
-  if (stdlib == NULL) stdlib = OCAML_STDLIB_DIR;
-  return stdlib;
+  if (path[0] == '.') {
+    if (path[1] == '\0') {
+      /* path is exactly "." => return root */
+      return caml_stat_strdup_os(root);
+    } else if (Is_separator(path[1])) {
+      /* path is exactly "./" or begins "./". In both cases, replace the "."
+         with root */
+      return caml_stat_strconcat_os(2, root, (path + 1));
+    } else if (path[1] == '.' && (path[2] == '\0' || Is_separator(path[2]))) {
+      /* path is either exactly ".." or begins "../" => prefix it with root
+         (which has no trailing separator) */
+      return caml_stat_strconcat_os(3, root, CAML_DIR_SEP, path);
+    } else {
+      /* path is not relative, but simply begins with a dot => return a copy */
+      return caml_stat_strdup_os(path);
+    }
+  } else {
+    /* path is not relative => return a copy */
+    return caml_stat_strdup_os(path);
+  }
 }
 
-CAMLexport char_os * caml_parse_ld_conf(void)
+CAMLexport char_os * caml_parse_ld_conf(const char_os * stdlib,
+                                        struct ext_table *table)
 {
-  const char_os * stdlib;
-  char_os * ldconfname, * wconfig, * p, * q;
+  const char_os * const locations[3] = {
+    caml_secure_getenv(T("OCAMLLIB")),
+    caml_secure_getenv(T("CAMLLIB")),
+    stdlib};
+  char_os * libroot, * ldconfname, * wconfig, * p, * q;
   char * config;
 #ifdef _WIN32
   struct _stati64 st;
 #else
   struct stat st;
 #endif
   int ldconf, nread;
+  size_t length = 0;
+  struct ext_table entries;
+
+  /* Use a temporary ext_table to hold the individually-allocated entries */
+  caml_ext_table_init(&entries, 8);
+  for (int i = 0; i < sizeof(locations) / sizeof(locations[0]); i++) {
+    if (locations[i] != NULL) {
+      libroot = caml_stat_strdup_os(locations[i]);
+      size_t libroot_length = strlen_os(libroot);
+      if (libroot_length > 1 && Is_separator(libroot[libroot_length - 1]))
+        libroot[libroot_length - 1] = '\0';
```

Comment:
If you want to handle the corner cases of :
- several slashes at the end of a path
- the user putting OCAMLLIB at the file system root 😂
- both of the above

```suggestion
      while (libroot_length > 0 && Is_separator(libroot[libroot_length - 1]))
        libroot[--libroot_length] = '\0';
```

Reply:
Technically-speaking (and more by luck than judgement) the previous code "correctly" preserved `//` as a root, where this version now treats it as `/`. A user with `OCAMLLIB=/` is already in a lot of trouble; the Cygwin user who ended up with `OCAMLLIB=//` and a host on the network named `ld.conf` was in even more trouble, and may come back and thank us for changing this...